sw/device/lib/crypto/impl/aes.c

```diff
@@ -0,0 +1,267 @@
+          memset(&padded_block->data[last_block_len + 1], 0x0, padding_len - 1);
+        }
+      }
+      break;
+    case kAesPaddingNull:
+      // No padding; input length must be an exact multiple of block size.
+      if (padding_len != 0) {
+        return kCryptoStatusBadArgs;
+      }
+      break;
+    default:
+      // TODO(#15591): Add any other padding modes that will be included in the
+      // final API.
+      return kCryptoStatusBadArgs;
+  }
+  return kCryptoStatusOK;
+}
+
+crypto_status_t otcrypto_aes(const crypto_blinded_key_t *key,
+                             crypto_uint8_buf_t iv,
+                             block_cipher_mode_t aes_mode,
+                             aes_operation_t aes_operation,
+                             crypto_const_uint8_buf_t cipher_input,
+                             aes_padding_t aes_padding,
+                             crypto_uint8_buf_t *cipher_output) {
+  // Check for NULL pointers in input pointers and data buffers.
+  if (key == NULL || cipher_output == NULL || iv.data == NULL ||
+      cipher_input.data == NULL || cipher_output->data == NULL) {
+    return kCryptoStatusBadArgs;
+  }
+
+  // Key integrity check.
+  if (blinded_key_integrity_check(key) != kHardenedBoolTrue) {
+    return kCryptoStatusBadArgs;
+  }
+
+  // Calculate the number of input blocks by rounding up to the nearest full
+  // block.
+  size_t input_nblocks =
+      (cipher_input.len + (kAesBlockNumBytes - 1)) / kAesBlockNumBytes;
+
+  // Check for bad lengths:
+  //   - IV must be exactly one block long.
+  //   - Input length must be nonzero.
+  //   - Output length must match number of input blocks.
+  if (iv.len != kAesBlockNumBytes || cipher_input.len == 0 ||
+      cipher_output->len != input_nblocks * kAesBlockNumBytes) {
+    return kCryptoStatusBadArgs;
+  }
+
+  // Load the IV.
+  aes_block_t aes_iv;
+  memcpy(aes_iv.data, iv.data, iv.len);
+
+  // Parse the AES key.
+  aes_key_t aes_key;
+  crypto_status_t err = construct_aes_key(key, aes_mode, &aes_key);
+  if (err != kCryptoStatusOK) {
+    return err;
+  }
+
+  // Start the operation (encryption or decryption).
+  aes_error_t aes_err;
+  switch (aes_operation) {
+    case kAesOperationEncrypt:
+      aes_err = aes_encrypt_begin(aes_key, &aes_iv);
+      if (aes_err != kAesOk) {
+        return kCryptoStatusInternalError;
+      }
+      break;
+    case kAesOperationDecrypt:
+      aes_err = aes_decrypt_begin(aes_key, &aes_iv);
+      if (aes_err != kAesOk) {
+        return kCryptoStatusInternalError;
+      }
+      break;
+    default:
+      return kCryptoStatusBadArgs;
+  }
+
+  // Perform the cipher operation for all full blocks (excluding last block).
+  // The input and output are offset by one, so if unrolled this loop would
+  // look like:
+  //   aes_update(NULL, input[0]);
+  //   aes_update(output[0], input[1]);
+  //   aes_update(output[1], input[2]);
+  //   ...
+  // See the AES driver for details.
+  aes_block_t block_in;
```

Comment:
unsure/curious: do we ever need to zeroize any stack allocated variables after use?

Reply:
probably fine bc this is executed in priviledged kernel mode and stack separate+protected from userland?

Reply:
Yes, the cryptolib runs within the TockOS kernel, so I think we don't need to clear anything on the stack. But I'll double-check that to be sure :slightly_smiling_face: